client/gc_client.go

```diff
@@ -0,0 +1,105 @@
+	WatchGCSafePointV2(ctx context.Context) (chan []*pdpb.SafePointEvent, error)
+}
+
+// UpdateGCSafePointV2 update gc safe point for the given keyspace.
+func (c *client) UpdateGCSafePointV2(ctx context.Context, keyspaceID uint32, safePoint uint64) (uint64, error) {
+	if span := opentracing.SpanFromContext(ctx); span != nil {
+		span = opentracing.StartSpan("pdclient.UpdateGCSafePointV2", opentracing.ChildOf(span.Context()))
+		defer span.Finish()
+	}
+	start := time.Now()
+	defer func() { cmdDurationUpdateGCSafePointV2.Observe(time.Since(start).Seconds()) }()
+
+	ctx, cancel := context.WithTimeout(ctx, c.option.timeout)
+	req := &pdpb.UpdateGCSafePointV2Request{
+		Header:     c.requestHeader(),
+		KeyspaceId: keyspaceID,
+		SafePoint:  safePoint,
+	}
+	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
+	resp, err := c.getClient().UpdateGCSafePointV2(ctx, req)
+	cancel()
+
+	if err = c.respForErr(cmdFailedDurationUpdateGCSafePointV2, start, err, resp.GetHeader()); err != nil {
+		return 0, err
+	}
+	return resp.GetNewSafePoint(), nil
+}
+
+// UpdateServiceSafePointV2 update service safe point for the given keyspace.
+func (c *client) UpdateServiceSafePointV2(ctx context.Context, keyspaceID uint32, serviceID string, ttl int64, safePoint uint64) (uint64, error) {
+	if span := opentracing.SpanFromContext(ctx); span != nil {
+		span = opentracing.StartSpan("pdclient.UpdateServiceSafePointV2", opentracing.ChildOf(span.Context()))
+		defer span.Finish()
+	}
+	start := time.Now()
+	defer func() { cmdDurationUpdateServiceSafePointV2.Observe(time.Since(start).Seconds()) }()
+
+	ctx, cancel := context.WithTimeout(ctx, c.option.timeout)
+	req := &pdpb.UpdateServiceSafePointV2Request{
+		Header:     c.requestHeader(),
+		KeyspaceId: keyspaceID,
+		ServiceId:  []byte(serviceID),
+		SafePoint:  safePoint,
+		Ttl:        ttl,
+	}
+	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
+	resp, err := c.getClient().UpdateServiceSafePointV2(ctx, req)
+	cancel()
+	if err = c.respForErr(cmdFailedDurationUpdateServiceSafePointV2, start, err, resp.GetHeader()); err != nil {
+		return 0, err
+	}
+	return resp.GetMinSafePoint(), nil
+}
+
+// WatchGCSafePointV2 watch gc safe point change.
+func (c *client) WatchGCSafePointV2(ctx context.Context) (chan []*pdpb.SafePointEvent, error) {
+	SafePointEventsChan := make(chan []*pdpb.SafePointEvent)
+	req := &pdpb.WatchGCSafePointV2Request{
+		Header: c.requestHeader(),
+	}
+	stream, err := c.getClient().WatchGCSafePointV2(ctx, req)
+	if err != nil {
+		close(SafePointEventsChan)
+		return nil, err
+	}
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				log.Error("[pd] panic in gc client `WatchGCSafePointV2`", zap.Any("error", r))
+				return
+			}
+		}()
+		for {
+			select {
+			case <-ctx.Done():
+				close(SafePointEventsChan)
+				return
+			default:
+				resp, err := stream.Recv()
+				if err != nil {
+					return
```

Comment:
need to close the channel

Reply:
Fix it like https://github.com/tikv/pd/pull/6494 .